packages/basic-crawler/src/internals/basic-crawler.ts

```diff
@@ -935,23 +938,63 @@ export class BasicCrawler<Context extends CrawlingContext = BasicCrawlingContext
      * Pushes data to the default crawler {@apilink Dataset} by calling {@apilink Dataset.pushData}.
      */
     async pushData(...args: Parameters<Dataset['pushData']>): Promise<void> {
-        const dataset = await Dataset.open(undefined, { config: this.config });
-        return dataset.pushData(...args);
+        if (this.dataset == null) {
+            this.dataset = await Dataset.open(undefined, { config: this.config });
+        }
+
+        return this.dataset.pushData(...args);
     }
 
     /**
-     * Retrieves the default crawler {@apilink Dataset} by calling {@apilink Dataset.open}.
+     * Retrieves the default crawler {@apilink Dataset}.
      */
-    async getDataset(): Promise<Dataset> {
-        return Dataset.open(undefined, { config: this.config });
+    getDataset(): Dataset {
+        return this.dataset;
     }
 
     /**
      * Retrieves data from the default crawler {@apilink Dataset} by calling {@apilink Dataset.getData}.
      */
     async getData(...args: Parameters<Dataset['getData']>): ReturnType<Dataset['getData']> {
-        const dataset = await this.getDataset();
-        return dataset.getData(...args);
+        return this.getDataset().getData(...args);
```

Comment:
Can't this throw now, since the dataset isn't guaranteed?

Reply:
Oh nvm, `_init` sets it... Then why also potentially init it in pushData?

Reply:
Some tests were using the helpers without calling `crawler.run()` which is rather specific for the tests, but still rather did it this way to stay safe.

Reply:
I will also add a validation for this into the `crawler.getDataset()` to be on the safe side.

Reply:
Maybe we should move it to the getDataset method?

Reply:
Move what? I wanted to make the method sync, so it cannot lazily create the dataset.

Reply:
Hmm...Fair enough

Reply:
On the other hand, with the new `crawler.exportData()` helper, it is not that important to have it sync, so maybe I'll just revert this and make it async again.

Reply:
reverted to the async version